// react-svg-pan-zoom/build-commonjs/constants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var MODE_IDLE = exports.MODE_IDLE = 'idle';
var MODE_PANNING = exports.MODE_PANNING = 'panning';
var MODE_ZOOMING = exports.MODE_ZOOMING = 'zooming';

var TOOL_AUTO = exports.TOOL_AUTO = 'auto';
var TOOL_NONE = exports.TOOL_NONE = 'none';
var TOOL_PAN = exports.TOOL_PAN = 'pan';
var TOOL_ZOOM_IN = exports.TOOL_ZOOM_IN = 'zoom-in';
var TOOL_ZOOM_OUT = exports.TOOL_ZOOM_OUT = 'zoom-out';

var POSITION_NONE = exports.POSITION_NONE = 'none';
var POSITION_TOP = exports.POSITION_TOP = 'top';
var POSITION_RIGHT = exports.POSITION_RIGHT = 'right';
var POSITION_BOTTOM = exports.POSITION_BOTTOM = 'bottom';
var POSITION_LEFT = exports.POSITION_LEFT = 'left';

var ACTION_ZOOM = exports.ACTION_ZOOM = 'zoom';
var ACTION_PAN = exports.ACTION_PAN = 'pan';

var ALIGN_CENTER = exports.ALIGN_CENTER = 'center';
var ALIGN_LEFT = exports.ALIGN_LEFT = 'left';
var ALIGN_RIGHT = exports.ALIGN_RIGHT = 'right';
var ALIGN_TOP = exports.ALIGN_TOP = 'top';
var ALIGN_BOTTOM = exports.ALIGN_BOTTOM = 'bottom';
